### PR TITLE
Add -f option for limiting snapshot frequency

### PR DESCRIPTION
**Pull request by DarwinAwardWinner**
Here is a potentially simpler alternative to #72 to solve #71. It just adds a `-f` option to `zfsnap snapshot` that prevents snapshot creation unless all existing snapshots with the specified prefix are older than the specified age. The idea is to run this in a frequently-executed cronjob. For example, I've set up the following daily cron job on my server:

```bash 
#!/bin/bash

# Sync root FS (which is not ZFS) to ZFS for snapshotting
ionice -c3 rsync -aAHX --quiet --one-file-system / /rpool/ubuntu/root-backup/

# Daily snapshot
/usr/local/sbin/zfsnap snapshot -p zfsnap-daily- -f 1d -a 2w -z -r rpool
# Weekly
/usr/local/sbin/zfsnap snapshot -p zfsnap-weekly- -f 1w -a 2m -z -r rpool
# Monthly
/usr/local/sbin/zfsnap snapshot -p zfsnap-monthly- -f 1m -a 2y -z -r rpool
# Yearly
/usr/local/sbin/zfsnap snapshot -p zfsnap-yearly- -f 1y -a 10y -z -r rpool

# Clean up expired snapshots
for prefix in zfsnap-{daily,weekly,monthly,yearly}-; do
        /usr/local/sbin/zfsnap destroy -p "$prefix" -r rpool
done
```

Another possible use case would be a server that is not expected to run continuously at all times. In this case, you could put something like the above in an hourly cron job, to ensure that snapshots will not be skipped simply because the server is never running at the appropriate time.

### CODE REVIEW

**Comment by DarwinAwardWinner** [comment]
I think I actually prefer this approach after having tried both. The main reason is that it lets me put my entire snapshot config in a single file.

**Comment by Alexqw** [comment]
Interesting idea. I'm still on vacation for another week, but I'll continue to digest this and look at it more closely when I'm back in work mode. in general though, I like the approach.

Thanks for your work on this.

---Alex

**Comment by rsyring** [comment]
For what it's worth, I really like this approach and would like to see it used.  I'm using ZFS on a laptop and run into issues with simple cron based "rotation" scripts that assume a machine is always up.  Being able to run a single cron script like the one in the original post for this PR makes a lot of sense to me.

@Alexqw hope you are enjoying your vacation.  :)

**Comment by graudeejs** [comment]
@rsyring, I recommend using anacron instead of cron

**Comment by Alexqw** [comment]
@DarwinAwardWinner First off, thanks for the patch and idea. I think this is a great way to handle this problem.

After reviewing the code, I have a few thoughts/suggestions.
* zfs lists snapshots in order, so we only have to check the most recent snapshot (meeting the prefix criteria) to determine if a snapshot should be performed.
* prefixes will be a problem. `destroy` uses a list of them to search for, `snapshot` uses a single prefix to prepend to the snapshot name. In this scenario, we're using prefix to do both. This is fine, unless users want to be able to search for any of a number of prefixes, but use one one when applying. I think this is an uncommon/unlikely scenario. But it's worth bringing up, as I don't see a way to handle that without breaking our syntax, which is a bad idea IMO. If we only allow filtering for and applying one (and the same) prefix, then I think this is doable.
* -f I think is the wrong letter to use. It's close to -F (which is force delete). I'm not terribly picky about what else is used; it's just better to stay away from current letters. We have a lot of open letters to choose from. Maybe -C for conditional?
* What do you think about a new subcommand for this, rather than adding the functionality to snapshot? I'm not sure if it's needed, but it keeps coming to mind for me. It might make it easier to integrate, and just have it call `snapshot` once it's determined if a snapshot is warranted.
* Multiple "-f"s could be specified. If a snapshot matches, then none of the options between the matching -f and the next pool/filesystem would be parsed. For example:
```
zfsnap maybeshot -p 'hourly-' -a 1d -f 1d -p 'daily-' -a 1w -f 1w -p 'monthly-' -a 1y -f 1y -p 'yearly-' -a forever pool/fs
```

What are you thoughts?

---Alex

**Comment by DarwinAwardWinner** [comment]
> zfs lists snapshots in order, so we only have to check the most recent snapshot (meeting the prefix criteria) to determine if a snapshot should be performed.

It looks like zfs lists snapshots with the latest ones last, so I guess the code could quickly go through all the snapshots, saving only the last one that matches, and then just do the date math on that one. But while it might be slightly more time-efficient, it wouldn't really simplify the code much, and it would add an assumption on the ordering of snapshots in the output `zfs list`. Does zfs explicitly guarantee the order in which it lists snapshots?

> prefixes will be a problem...

So you're saying there should be a way to, for example, avoid creating a daily snapshot if any of daily/weekly/monthly/yearly snapshots was created within the past day? I considered that, but decided it didn't make sense to implement within the existing snapshot command. This would need to be implemented as part of a separate command, I think. In any case, while I agree it's slightly suboptimal, the worst case here is that a few extra snapshots are created, but at the same time it has the desirable property that each prefix is handled independently of the others. 

> -f I think is the wrong letter to use

What about `-m` for "max frequency"?

> What do you think about a new subcommand for this, rather than adding the functionality to snapshot?

Maybe a new sub-command that just determines whether a new snapshot is necessary based on the criteria given, and exits with 0 or 1 appropriately? E.g., to make a daily snapshot only if a daily, weekly, monthly, or yearly snapshot has not been made in the past day:

    zfsnap need-snapshot -m 1d -p 'daily-' -p 'weekly-' -p 'monthly-' -p 'yearly-' pool/fs
    RETCODE="$?"
 then
        zfsnap snapshot -p 'daily-' -a '1w' pool/fs
    fi

The problem with that would be that if an error occurs, `zfsnap need-snapshot` would exit with a non-zero code and no snapshot would be created. 


For the multiple option suggestion, I guess you're suggesting it should work similarly to the `-p` and `-P` options? Except you don't want it to apply filesystems that follow the option, you want it to apply to all following prefixes? I find your example command quite hard to parse. It seems like some arguments apply to the following prefix while others apply to the preceding one. I personally don't use the "multiple prefixes in a single command" feature at all. I just do as shown above in my example, running a separate command for each prefix. For me, at least, that's easier to understand, even if it might take a bit longer to run.

**Comment by DarwinAwardWinner** [comment]
> zfs lists snapshots in order, so we only have to check the most recent snapshot (meeting the prefix criteria) to determine if a snapshot should be performed.

It looks like zfs lists snapshots with the latest ones last, so I guess the code could quickly go through all the snapshots, saving only the last one that matches, and then just do the date math on that one. But while it might be slightly more time-efficient, it wouldn't really simplify the code much, and it would add an assumption on the ordering of snapshots in the output `zfs list`. Does zfs explicitly guarantee the order in which it lists snapshots?

> prefixes will be a problem...

So you're saying there should be a way to, for example, avoid creating a daily snapshot if any of daily/weekly/monthly/yearly snapshots was created within the past day? I considered that, but decided it didn't make sense to implement within the existing snapshot command. This would need to be implemented as part of a separate command, I think. In any case, while I agree it's slightly suboptimal, the worst case here is that a few extra snapshots are created, but at the same time it has the desirable property that each prefix is handled independently of the others. 

> -f I think is the wrong letter to use

What about `-m` for "max frequency"?

> What do you think about a new subcommand for this, rather than adding the functionality to snapshot?

As I mentioned above, I think certain features, like searching multiple prefixes for a recent snapshot, would probably necessitate a separate command.

> Multiple "-f"s could be specified. If a snapshot matches, then none of the options between the matching -f and the next pool/filesystem would be parsed.

I find the example command hard to parse mentally, since some options, like `-a`, apply to the previous prefix, while others, like `-f`, apply to the following prefixes; and also, because there is a very complex order dependency among the options and arguments: prefixes apply to the following pool/fs, while `-f` and `-a` options apply to prefixes. (And this also requires a major change in the option parsing to accommodate.) I think you could achieve a similar result by having each `-f` apply to the following pool/fs, just like `-p`:

    zfsnap maybeshot \
        -p 'hourly-' -a 1d -f 1h pool/fs \
        -p 'daily-' -a 1w -f 1d pool/fs \
        -p 'weekly-' -a 1m -f 1w pool/fs \
        -p 'monthly-' -a 1y -f 1m pool/fs \
        -p 'yearly-' -a forever -f 1y pool/fs

I personally don't use the "multiple prefixes/pools in a single command" feature at all. I just do as shown above in my original example, running a separate command for each prefix and/or pool. For me, at least, that's easier to understand, even if it might take a bit longer to run.

**Comment by DarwinAwardWinner** [comment]
Just to clarify, I have no idea if my current code implements the behavior described for the example command in my previous comment, since I haven't looked at how zfsnap handles multiple pool/fs arguments. I'm merely suggesting that that is how it could work.

**Comment by DarwinAwardWinner** [comment]
By the way, in my testing I have confirmed that the frequency option requires a "fudge factor" in order to work correctly. For example, if you do your daily snapshots with `-f 1d` or `-f 24h`, and you have a cronjob that runs every hour, you will end up with a snapshot every 25 hours, because on the run exactly 24 hours after a snapshot, it will still consider it within the 24 hour period and decline to make a snapshot. So my current setup uses 59M instead of 1h, 23h59M instead of 1d, etc., which is obviously not ideal.

**Comment by davidblewett** [comment]
> zfs lists snapshots in order, so we only have to check the most recent snapshot (meeting the prefix criteria) to determine if a snapshot should be performed.
> 
> It looks like zfs lists snapshots with the latest ones last, so I guess the code could quickly go through all the snapshots, saving only the last one that matches, and then just do the date math on that one. But while it might be slightly more time-efficient, it wouldn't really simplify the code much, and it would add an assumption on the ordering of snapshots in the output zfs list. Does zfs explicitly guarantee the order in which it lists snapshots?

You can force specific ordering using `-s` / `-S` (This is on FreeBSD 10.1, but ZFS on Linux appears to support this as well; don't have other platforms to test):
```
pharaon% zfs list -o name,creation -s creation
NAME                            CREATION
data                            Fri Aug  9 19:41 2013
data/photos                     Fri Aug  9 20:48 2013
data/music_archive              Sat Aug 10  2:15 2013
data/playlists                  Sat Aug 10  4:39 2013
data/zroot                      Sat Aug 10  5:20 2013
data/ebooks                     Sat Aug 10  5:30 2013
data/ipad                       Sat Aug 10  5:33 2013
data/audiobooks                 Sat Aug 10  5:34 2013
data/plex                       Mon Nov  4 13:46 2013
data/pgsql                      Wed Jan  1 17:50 2014
data/unifi                      Sat Jan 24 20:29 2015
data/circus                     Mon Jan 26 21:02 2015
data/minecraft                  Wed Aug 19 13:01 2015
data/transcode                  Wed Oct  7 19:25 2015
data/syncthing                  Wed Nov 11 11:58 2015
data/movies                     Tue Dec  1 15:09 2015
data/tv                         Tue Dec  1 21:21 2015
data/live_music                 Sat Apr  9 15:10 2016
```

```
pharaon% zfs list -o name,creation -S creation
NAME                            CREATION
data/live_music                 Sat Apr  9 15:10 2016
data/tv                         Tue Dec  1 21:21 2015
data/movies                     Tue Dec  1 15:09 2015
data/syncthing                  Wed Nov 11 11:58 2015
data/transcode                  Wed Oct  7 19:25 2015
data/minecraft                  Wed Aug 19 13:01 2015
data/circus                     Mon Jan 26 21:02 2015
data/unifi                      Sat Jan 24 20:29 2015
data/pgsql                      Wed Jan  1 17:50 2014
data/plex                       Mon Nov  4 13:46 2013
data/audiobooks                 Sat Aug 10  5:34 2013
data/ipad                       Sat Aug 10  5:33 2013
data/ebooks                     Sat Aug 10  5:30 2013
data/zroot                      Sat Aug 10  5:20 2013
data/playlists                  Sat Aug 10  4:39 2013
data/music_archive              Sat Aug 10  2:15 2013
data/photos                     Fri Aug  9 20:48 2013
data                            Fri Aug  9 19:41 2013
```
